Source/GMCAbilitySystem/Private/Components/GMCAbilityComponent.cpp

```diff
@@ -1158,6 +1143,132 @@ void UGMC_AbilitySystemComponent::InitializeStartingAbilities()
 	}
 }
 
+bool UGMC_AbilitySystemComponent::ProcessAbilityOperation(
+	const TGMASBoundQueueOperation<UGMCAbility, FGMCAbilityData>& Operation, bool bFromMovementTick)
+{
+	EGMASBoundQueueOperationType OperationType = Operation.GetOperationType();
+	if (OperationType == EGMASBoundQueueOperationType::Activate)
+	{
+		TryActivateAbilitiesByInputTag(Operation.GetTag(), Operation.Payload.ActionInput, bFromMovementTick);
+		return true;
+	}
+
+	if (OperationType == EGMASBoundQueueOperationType::Cancel)
+	{
+		EndAbilitiesByTag(Operation.GetTag());
+		if (Operation.ItemClass)
+		{
+			EndAbilitiesByClass(Operation.ItemClass);
+		}
+		return true;
+	}
+	
+	UE_LOG(LogGMCAbilitySystem, Warning, TEXT("Received ability operation with invalid operation type %s for %s!"),
+		*UEnum::GetValueAsString(OperationType), *Operation.GetTag().ToString())
+	return false;
+}
+
+UGMCAbilityEffect* UGMC_AbilitySystemComponent::ProcessEffectOperation(
+	const TGMASBoundQueueOperation<UGMCAbilityEffect, FGMCAbilityEffectData>& Operation)
+{
+	EGMASBoundQueueOperationType OperationType = Operation.GetOperationType();
+
+	if (OperationType == EGMASBoundQueueOperationType::Add)
+	{
+		if (Operation.ItemClass == nullptr)
+		{
+			UE_LOG(LogGMCAbilitySystem, Error, TEXT("[%20s] %s attempted to process an add effect operation with no set class!"),
+				*GetNetRoleAsString(GetOwnerRole()), *GetOwner()->GetName())
+		}
+		
+		UGMCAbilityEffect* Effect = DuplicateObject(Operation.ItemClass->GetDefaultObject<UGMCAbilityEffect>(), this);
```

Comment:
Any reason why `DuplicateObject` is used over `NewObject` in this case? Since the duplicated object is the CDO, it's basically the same as just calling `NewObject` with the item class no?

Reply:
No reason really, other than "it was done that way in other bits of the code which I didn't write." I was just trying to match the existing behavior for effects in case there _was_ a side effect I wasn't aware of. I have zero objection to switching it back to NewObject.

Reply:
Annoyingly it won't let me reply to the grace period comment, so I'm replying here. 

I used 1 second because the existing "outside" behavior used that grace period; in case anyone was using that code, I didn't want to change the value on them by changing what was happening under the hood. I do think we should expose it, and if no one is using the previous behavior I agree it would be good to have a shorter interval.